### PR TITLE
Pipes - Lindsey and Sairagul - Scrabble

### PR DESCRIPTION
**Pull request by sairagula**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |  So that we didn't have to create an instance of a Score class for every method we needed. |
|   Describe an example of a test you wrote for a _nominal case_. |   We checked draw_tiles (tile_bag) method to draw 2 tiles from it, so  in tiles array 5 tiles should be at the first hand, before drawing, so that max tiles will be 7  	|
|   Describe an example of a test you wrote for an _edge case_. |   We tested if the player could play the same word again if he has already won it	|
|   What was your final test coverage percentage?	|   98.54	|
|   Is there a specific piece of code you'd like feedback on?	|  We need feedback on the parts of the code that have been commented out. Because we couldn't figure out why it is not working  with the remaining code and doesn't pass tests. At the same time it works in another computer with the same code and passes all tests	|

### CODE REVIEW

File: lib/score.rb [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    end
+
+    # calculates the score of a given word
+    def self.score(word)
+	  word = word.upcase
+      score_chart =
+      {
+        1 => ["A", "E", "I", "O", "U", "L", "N", "R", "S", "T"],
+        2 => ["D", "G"],
+        3 => ["B", "C", "M", "P"],
```
Comment (droberts-sea) [inline lib/score.rb:15]: This data structure certainly works, but it means you need a double loop below. It might be a little cleaner to have the map go the other way, that is, from letters to score. Then finding a letter's score is as concise as `score_chart[letter]`.

File: lib/score.rb [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        end
+      end
+      return longest_string
+    end
+
+
+     # returns true if word2 is better choice.
+     def self.is_better_choice(word1, word2)
+       score1 = score(word1)
+       score2 = score(word2)
```
Comment (droberts-sea) [inline lib/score.rb:91]: Good work breaking this logic out into its own method.

File: spec/score_spec.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    it " Check for the 7 letter word " do
+      my_class = Scrabble::Score
+      my_class.score("BUBBLES").must_equal 63
+    end
+
+	it " Check for case sensitivity " do
+      my_class = Scrabble::Score
+      my_class.score("bubbles").must_equal 63
+    end
```
Comment (droberts-sea) [inline spec/score_spec.rb:30]: What about an empty word, or an 8-letter word?

File: spec/score_spec.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    it " if tie, 7 letter word should be returned" do
+      my_class = Scrabble::Score
+      my_class.highest_score_from_array(["DOLLAR", "HALL","LETTERS"]).must_equal "LETTERS"
+    end
+
+    it " If 2 words with 7 letters in tie, first word should be returned" do
+      my_class = Scrabble::Score
+      my_class.highest_score_from_array(["QUICKLY", "SQUEEZE","LETTERS"]).must_equal "QUICKLY"
+    end
```
Comment (droberts-sea) [inline spec/score_spec.rb:53]: What happens if you run the method with an empty array, or an array with only one word?

What about when you pass in an array with a 6-letter and 7-letter word with the same length? For example, "ZZZZZZ" and "AAAAAAW"?

File: lib/player.rb [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      @plays.each do |i|
+        total_so_far += Scrabble::Score.score(i)
+      end
+      return total_so_far
+    end
+
+
+    def won?
+      total_score > 100 ? true : false
+    end
```
Comment (droberts-sea) [inline lib/player.rb:53]: You don't need the ternary operator here - because `total_score > 100` already evaluates to a boolean, you can just return that.

File: spec/player_spec.rb [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+      value = player.play('hello')
+
+      player.plays.length.must_equal 1
+      value.must_equal 8
+    end
+
+    it "plays after a player has won" do
+      name = 'John'
+      player = Scrabble::Player.new(name)
+      player.play('bubbles')
```
Comment (droberts-sea) [inline spec/player_spec.rb:70]: Good test case! This is an important behavior for the game, but it would be an easy one to miss.

File: spec/tilebag_spec.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+    it "Should create a collection of default tiles" do
+      tilebag = Scrabble::TileBag.new
+      tilebag.tiles_remaining.must_equal 98
+    end
+  end
+
+  describe "#draw_tiles method" do
+    it "Should make sure a max of 7 tiles drawn" do
+  	tilebag = Scrabble::TileBag.new
```
Comment (droberts-sea) [inline spec/tilebag_spec.rb:18]: You're missing a couple of interesting edge cases here:
- What happens if you attempt to draw 0 tiles?
- Are you able to draw the bag all the way down to 0 tiles remaining?
- What happens when you attempt to draw more tiles than are in the bag?

**Comment by droberts-sea** [comment]
# Scrabble
## What We're Looking For

|  Feature  |   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions  |  yes  |
|   Both Teammates contributed to the codebase	|  we have discussed this offline  |
|   Regular Commits with meaningful commit messages  |  Commit frequency looks good, but some of your commit messages could use some work. Make sure they provide a good description of what changed and why.  |
|  **Scoring class**	|   	|
|  `score` method	|  yes  |
|   Uses appropriate data structure to store the letter score	|  see inline comment  |
|   Appropriately handles edge cases	|  yes  |
|   Tests `score` edge cases (one-letter word, 7 letter word)	|  see inline comment  |
|  `highest_scoring_word` method	|  yes  |
|   Appropriately handles edge cases	|  yes  |
|   Tests `highest_scoring_word` edge cases (ties, no words, no array)	|  see inline comment  |
|  **Player class**	|   	|
|  Uses an array to keep track of words played	|  yes  |
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|  yes  |
|  Returns `true` or `false` from `won?` method |  yes  |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |  yes  |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	|  yes  |
|  `draw_tiles` method uses the parameter to draw a set of random tiles |  yes  |
| `tiles_remaining` method returns a count	|  yes  |
|  Tests `draw_tiles` edge cases for different parameter inputs	|  see inline comment  |

Good work overall, especially on product code. There are a few places where you've got holes in your test coverage, which I've tried to outline below. The pattern I'm seeing is that you should make sure to explore both sides of any edge cases. For example, in `TileBag#draw_tiles`, it's important to consider what happens if you attempt to draw more tiles than are in the bag (this should fail), and to check that you can draw the last tile in the bag (this should work). You can look at your code and say "of course that would work", but until you test it you can't know for sure.

Becoming a master tester is not easy, and this is the first project where you've had to build tests around complex logic, so don't worry if it's not yet where we want it to be. We'll have plenty of time to practice going forward. Do make sure to spend time thinking about it for Hotel, particularly around how ranges of dates interact, and let one of us know if you want us to check your work.
